source/game/kart/KartDynamics.cc

```diff
@@ -7,34 +7,59 @@ KartDynamics::KartDynamics() {
 }
 
 void KartDynamics::init() {
-    m_pos.setZero();
-    m_speed.setZero();
-    m_fullRot = EGG::Quatf(1.0f, EGG::Vector3f::zero);
+    m_speedNorm = 0.0f;
+    m_gravity = -1.0f;
     m_top = EGG::Vector3f::ey;
+    m_forceUpright = true;
+    m_noGravity = false;
 }
 
-const EGG::Vector3f &KartDynamics::getPos() const {
+void KartDynamics::resetInternalVelocity() {
+    m_velocity.setZero();
+}
+
+void KartDynamics::calc(f32 dt, f32 /*maxSpeed*/, bool /*air*/) {
+    if (!m_noGravity) {
+        m_totalForce.y += m_gravity;
+    }
+
+    m_acceleration0 = m_totalForce;
+    m_externalVel += m_acceleration0 * dt;
+    m_externalVel *= 0.998f;
+    m_velocity =
+            m_externalVel * dt + m_internalVelocity + m_movingObjectVelocity + m_movingRoadVelocity;
+    m_speedNorm = m_velocity.normalise();
+    m_velocity *= m_speedNorm;
```

Comment:
```suggestion
    m_speedNorm = std::min(m_velocity.normalise(), maxSpeed);
    m_velocity *= m_speedNorm;
```

In the future, we can handle `maxSpeed`'s change due to the scale Z component beforehand with: 
```cpp
if (m_scale.z > 1.0f) {
    maxSpeed *= m_scale.z;
}
```